<a href="https://colab.research.google.com/github/Hany-Kim/DataPreprocessing_and_SearchLibrary_KOSA/blob/main/DAY_24_05_22/numpyDay3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 2장. 데이터프레임과 시리즈

## 1절. 판다스 패키지

In [1]:
import pandas as pd

In [6]:
pd.options.display.max_columns=999

## 2절. 데이터프레임만들기

### 딕셔너리를 이용해서 데이터프레임 만들기

딕셔너리 값(value)가 리스트인 경우

In [7]:
df = pd.DataFrame(data={'col1':[1,2], 'col2':[3,4]})
df

col1  col2
0     1     3
1     2     4

각 행의 정보가 딕셔너리 형태로 주어지는 경우\
[{key:val}, {key:val}, ...]

In [11]:
df = pd.DataFrame(data=[{'col1':1, 'col2':3},
                        {'col1':2, 'col2':4}])
df

col1  col2
0     1     3
1     2     4

각 행의 정보가 딕셔너리 형태로 주어지지만,\
열의 원소갯수가 모자르다면 NaN으로 채워준다.

In [13]:
df = pd.DataFrame(data=[{'col1':1, 'col2':3},
                        {'col1':2, 'col2':4},
                        {'col1':5}])
df

col1  col2
0     1   3.0
1     2   4.0
2     5   NaN

### 리스트를 이용해 데이터프레임 만들기

2개 이상의 리스트를 이용해 데이터프레임 생성하는 법

In [170]:
L1 = [1,2,3,4,5]
L2 = [6,7,8,9,10]

In [16]:
df = pd.DataFrame(data={'col1':L1, 'col2':L2})
df

col1  col2
0     1     6
1     2     7
2     3     8
3     4     9
4     5    10

In [17]:
import numpy as np

c_[]로 리스트를 열 단위로 연결 후\
columns로 열 이름 지정, 지정하지 않으면 0열부터 순서대로 자동생성

In [168]:
df = pd.DataFrame(np.c_[L1, L2],
                  columns=['col1','col2'])
df

col1  col2
0     1     6
1     2     7
2     3     8
3     4     9
4     5    10

In [169]:
df = pd.DataFrame(np.c_[L1, L2])
df

0   1
0  1   6
1  2   7
2  3   8
3  4   9
4  5  10

In [174]:
df = pd.DataFrame(np.r_[L1, L2],
                  index=['row1','row2','row3','row4','row5',
                         'row6','row7','row8','row9','row10'])
df

0
row1    1
row2    2
row3    3
row4    4
row5    5
row6    6
row7    7
row8    8
row9    9
row10  10

L1, L2가 리스트의 리스트(2차원 리스트)인 경우

In [19]:
L1 = [[1,2,3,4,5]]
L2 = [[6,7,8,9,10]]

c_[]를 사용하면 1개의 행으로 표현된다.

In [20]:
df = pd.DataFrame(np.c_[L1,L2])
df

0  1  2  3  4  5  6  7  8   9
0  1  2  3  4  5  6  7  8  9  10

r_[]를 사용하면 행단위로 리스트를 쌓는다.

In [21]:
df = pd.DataFrame(np.r_[L1,L2])
df

0  1  2  3   4
0  1  2  3  4   5
1  6  7  8  9  10

In [177]:
L1 = [[1,2,3],[6,7,8]]
L2 = [[8,9,10],[11,12,13]]

In [178]:
df = pd.DataFrame(np.c_[L1,L2])
df

0  1  2   3   4   5
0  1  2  3   8   9  10
1  6  7  8  11  12  13

In [179]:
df = pd.DataFrame(np.r_[L1,L2])
df

0   1   2
0   1   2   3
1   6   7   8
2   8   9  10
3  11  12  13

### iris 데이터셋
붓꽃의 3종류
> setosa(세토사) | versicolor(버시컬러) | virginica(버지니카)
* 각 종류 별로 50개씩 Data를 모아둠
* 각 Data의 정보 종류로는
    * Sepal length & width => 꽃받침 길이 & 너비
    * Petal length & width => 꽃잎 길이 & 너비
    * Species => 종

#### Scikit-learn패키지의 iris 데이터셋

In [22]:
from sklearn import datasets

* iris 데이터는 딕셔너리 형태
    * key값 = {data, target, target_names, frame,  DESCR, feature_names, filename, data_module, ...}
    * data => 독립변수 : 어떤 것에 영향을 주는 변수
    * target => 종속변수 : 어떤 것으로부터 영향을 받는 변수

In [184]:
iris_dic = datasets.load_iris()
iris_dic

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

data와 target을 이용해 x(독립변수), y(종속변수)를 DataFrame으로 만듬

In [185]:
x = pd.DataFrame(iris_dic.data, columns=iris_dic.feature_names)
y = pd.DataFrame(iris_dic.target_names[iris_dic.target], columns=['species'])

In [186]:
print(x)
print(y)

     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]
       species
0       setosa
1       se

위 코드를 딕셔너리를 참조하는 형태로 []를 사용해 같은 결과값을 출력

In [26]:
x = pd.DataFrame(iris_dic['data'], columns=iris_dic['feature_names'])
y = pd.DataFrame(iris_dic['target_names'][iris_dic['target']], columns=['species'])

concat()으로 x와 y를 붙여서 하나의 DataFrame으로 생성\
axis=1(열,오른쪽)에 붙임

In [189]:
iris = pd.concat([x,y], axis=1)
iris.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

  species  
0  setosa  
1  setosa  
2  setosa  
3  setosa  
4  setosa

concat()으로 x와 y를 붙여서 하나의 DataFrame으로 생성\
axis=0(행,아래)에 붙임\
원소가 비어있는 곳은 NaN

In [188]:
iris = pd.concat([x,y], axis=0)
iris

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                  5.1               3.5                1.4               0.2   
1                  4.9               3.0                1.4               0.2   
2                  4.7               3.2                1.3               0.2   
3                  4.6               3.1                1.5               0.2   
4                  5.0               3.6                1.4               0.2   
..                 ...               ...                ...               ...   
145                NaN               NaN                NaN               NaN   
146                NaN               NaN                NaN               NaN   
147                NaN               NaN                NaN               NaN   
148                NaN               NaN                NaN               NaN   
149                NaN               NaN                NaN               NaN   

       species  
0          NaN  
1          NaN  
2          NaN  
3          NaN  
4          NaN  
..         ...  
145  virginica  
146  virginica  
147  virginica  
148  virginica  
149  virginica  

[300 rows x 5 columns]

species에 이름이 아닌 숫자 값을 가지게 함\
setosa => 0\
versicolor => 1\
viginica => 2

In [28]:
y = pd.DataFrame(iris_dic.target, columns=['species'])
iris = pd.concat([x,y], axis=1)

In [29]:
iris.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   species  
0        0  
1        0  
2        0  
3        0  
4        0

In [30]:
iris.columns = ['sepal_length', 'sepal_width', 'petal_length', 'petal_width', 'species']

In [31]:
iris.head()

sepal_length  sepal_width  petal_length  petal_width  species
0           5.1          3.5           1.4          0.2        0
1           4.9          3.0           1.4          0.2        0
2           4.7          3.2           1.3          0.2        0
3           4.6          3.1           1.5          0.2        0
4           5.0          3.6           1.4          0.2        0

#### statsmodels패키지의 iris 데이터셋

In [32]:
import statsmodels.api as sm

get_rdataset()함수로 iris데이터를 불러옴\
get_rdataset()는 R에서 사용하는 dataset을 불러오는데 사용

In [33]:
iris_data = sm.datasets.get_rdataset('iris', package='datasets')

In [34]:
type(iris_data)

statsmodels.datasets.utils.Dataset

데이터 셋의 이름

In [35]:
iris_data.title

"Edgar Anderson's Iris Data"

In [36]:
print(iris_data.__doc__)

.. container::

   .. container::

      ==== ===============
      iris R Documentation
      ==== ===============

      .. rubric:: Edgar Anderson's Iris Data
         :name: edgar-andersons-iris-data

      .. rubric:: Description
         :name: description

      This famous (Fisher's or Anderson's) iris data set gives the
      measurements in centimeters of the variables sepal length and
      width and petal length and width, respectively, for 50 flowers
      from each of 3 species of iris. The species are *Iris setosa*,
      *versicolor*, and *virginica*.

      .. rubric:: Usage
         :name: usage

      ::

         iris
         iris3

      .. rubric:: Format
         :name: format

      ``iris`` is a data frame with 150 cases (rows) and 5 variables
      (columns) named ``Sepal.Length``, ``Sepal.Width``,
      ``Petal.Length``, ``Petal.Width``, and ``Species``.

      ``iris3`` gives the same data arranged as a 3-dimensional array of
      size 50 by 4 by 3, as onc

In [37]:
iris = iris_data.data
iris.head()

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width Species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

In [38]:
import seaborn as sns

In [39]:
iris = sns.load_dataset('iris')
iris.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

In [40]:
iris.to_csv('iris.csv', sep=',', index=False)

In [41]:
iris.to_csv('iris.csv.gz', sep=',', index=False, compression='infer')

In [42]:
iris = pd.read_csv('iris.csv')
iris.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

In [43]:
iris = pd.read_csv('iris.csv', skiprows=[0,2])
iris.head()

5.1  3.5  1.4  0.2  setosa
0  4.7  3.2  1.3  0.2  setosa
1  4.6  3.1  1.5  0.2  setosa
2  5.0  3.6  1.4  0.2  setosa
3  5.4  3.9  1.7  0.4  setosa
4  4.6  3.4  1.4  0.3  setosa

In [44]:
iris = pd.read_csv('iris.csv', comment='#')

In [45]:
iris = sns.load_dataset('iris')

In [46]:
iris.columns

Index(['sepal_length', 'sepal_width', 'petal_length', 'petal_width',
       'species'],
      dtype='object')

In [47]:
iris.columns = ['sl','sw','pl','pw','sp']
iris.head()

sl   sw   pl   pw      sp
0  5.1  3.5  1.4  0.2  setosa
1  4.9  3.0  1.4  0.2  setosa
2  4.7  3.2  1.3  0.2  setosa
3  4.6  3.1  1.5  0.2  setosa
4  5.0  3.6  1.4  0.2  setosa

In [48]:
print(list(iris.index))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149]


In [49]:
iris.index = range(150, 300)
iris.head()

sl   sw   pl   pw      sp
150  5.1  3.5  1.4  0.2  setosa
151  4.9  3.0  1.4  0.2  setosa
152  4.7  3.2  1.3  0.2  setosa
153  4.6  3.1  1.5  0.2  setosa
154  5.0  3.6  1.4  0.2  setosa

In [50]:
iris = pd.read_csv('iris.csv')

In [51]:
iris.columns = [['sepal', 'sepal', 'petal', 'petal', 'species'],
                ['legnth', 'width', 'length', 'width', 'species']]

In [52]:
iris.columns.names = ['sps','lw']

In [53]:
iris.head()

sps  sepal        petal       species
lw  legnth width length width species
0      5.1   3.5    1.4   0.2  setosa
1      4.9   3.0    1.4   0.2  setosa
2      4.7   3.2    1.3   0.2  setosa
3      4.6   3.1    1.5   0.2  setosa
4      5.0   3.6    1.4   0.2  setosa

In [54]:
iris.index = [['setosa' for i in range(50)] +
              ['versicolor' for i in range(50)] +
              ['virginica' for i in range(50)], list(range(150))]

In [55]:
iris.index.names = ['sp', 'rownum']

In [56]:
iris.head()

sps            sepal        petal       species
lw            legnth width length width species
sp     rownum                                  
setosa 0         5.1   3.5    1.4   0.2  setosa
       1         4.9   3.0    1.4   0.2  setosa
       2         4.7   3.2    1.3   0.2  setosa
       3         4.6   3.1    1.5   0.2  setosa
       4         5.0   3.6    1.4   0.2  setosa

In [65]:
iris.loc[('setosa',[0,1,2,3]), :]

TypeError: unhashable type: 'list'

In [57]:
iris = sns.load_dataset('iris')

In [58]:
iris.sepal_length

0      5.1
1      4.9
2      4.7
3      4.6
4      5.0
      ... 
145    6.7
146    6.3
147    6.5
148    6.2
149    5.9
Name: sepal_length, Length: 150, dtype: float64

In [59]:
iris['sepal_length']

0      5.1
1      4.9
2      4.7
3      4.6
4      5.0
      ... 
145    6.7
146    6.3
147    6.5
148    6.2
149    5.9
Name: sepal_length, Length: 150, dtype: float64

In [60]:
iris.loc[0]

sepal_length       5.1
sepal_width        3.5
petal_length       1.4
petal_width        0.2
species         setosa
Name: 0, dtype: object

In [61]:
iris.loc[0:5]

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa
5           5.4          3.9           1.7          0.4  setosa

In [62]:
iris.loc['sepal_length']

KeyError: 'sepal_length'

In [63]:
iris.loc['sepal_length':'sepal_width']

Empty DataFrame
Columns: [sepal_length, sepal_width, petal_length, petal_width, species]
Index: []

In [64]:
iris.loc[:,'sepal_length']

0      5.1
1      4.9
2      4.7
3      4.6
4      5.0
      ... 
145    6.7
146    6.3
147    6.5
148    6.2
149    5.9
Name: sepal_length, Length: 150, dtype: float64

In [66]:
iris.iloc[0:5]

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

In [67]:
iris.iloc[0:5, 0:3]

sepal_length  sepal_width  petal_length
0           5.1          3.5           1.4
1           4.9          3.0           1.4
2           4.7          3.2           1.3
3           4.6          3.1           1.5
4           5.0          3.6           1.4

In [68]:
iris.iloc[0:10:2, ::2]

sepal_length  petal_length species
0           5.1           1.4  setosa
2           4.7           1.3  setosa
4           5.0           1.4  setosa
6           4.6           1.4  setosa
8           4.4           1.4  setosa

In [69]:
iris_x = iris.iloc[:, :-1]
iris_y = iris.iloc[:, -1]
iris_x.head(2)

sepal_length  sepal_width  petal_length  petal_width
0           5.1          3.5           1.4          0.2
1           4.9          3.0           1.4          0.2

In [70]:
iris.loc[iris['species'] == 'versicolor'].head()

sepal_length  sepal_width  petal_length  petal_width     species
50           7.0          3.2           4.7          1.4  versicolor
51           6.4          3.2           4.5          1.5  versicolor
52           6.9          3.1           4.9          1.5  versicolor
53           5.5          2.3           4.0          1.3  versicolor
54           6.5          2.8           4.6          1.5  versicolor

In [71]:
iris.loc[iris['species'] == 'versicolor',
         ['sepal_length', 'sepal_width']].head()

sepal_length  sepal_width
50           7.0          3.2
51           6.4          3.2
52           6.9          3.1
53           5.5          2.3
54           6.5          2.8

In [72]:
iris.loc[(iris['species']=='versicolor') &
         (iris['sepal_length'] > 6.5)].head()

sepal_length  sepal_width  petal_length  petal_width     species
50           7.0          3.2           4.7          1.4  versicolor
52           6.9          3.1           4.9          1.5  versicolor
58           6.6          2.9           4.6          1.3  versicolor
65           6.7          3.1           4.4          1.4  versicolor
75           6.6          3.0           4.4          1.4  versicolor

In [73]:
df = pd.DataFrame({'X' : [1,2,3], 'Y':[6,7,8]}, index=['A','B','C'])

In [74]:
df.loc[df['Y']%2==0].loc[:, 'X'] = [10,30]

In [75]:
df

X  Y
A  1  6
B  2  7
C  3  8

In [76]:
df.loc[: 'X'].loc[df['Y']%2==0] = [10,30]

In [77]:
df

X   Y
A  10  30
B   2   7
C  10  30

In [78]:
df.loc[df['Y']%2==0, 'X'] = [10,30]

In [79]:
df

X   Y
A  10  30
B   2   7
C  30  30

In [83]:
df.at[df['Y']%2==0, 'X'] = [10,30]

InvalidIndexError: You can only assign a scalar value not a <class 'list'>

In [81]:
df

X   Y
A  10  30
B   2   7
C  30  30

In [101]:
iris = sns.load_dataset('iris')

In [102]:
iris.species

0         setosa
1         setosa
2         setosa
3         setosa
4         setosa
         ...    
145    virginica
146    virginica
147    virginica
148    virginica
149    virginica
Name: species, Length: 150, dtype: object

In [103]:
iris['new_col'] = 100
iris

sepal_length  sepal_width  petal_length  petal_width    species  new_col
0             5.1          3.5           1.4          0.2     setosa      100
1             4.9          3.0           1.4          0.2     setosa      100
2             4.7          3.2           1.3          0.2     setosa      100
3             4.6          3.1           1.5          0.2     setosa      100
4             5.0          3.6           1.4          0.2     setosa      100
..            ...          ...           ...          ...        ...      ...
145           6.7          3.0           5.2          2.3  virginica      100
146           6.3          2.5           5.0          1.9  virginica      100
147           6.5          3.0           5.2          2.0  virginica      100
148           6.2          3.4           5.4          2.3  virginica      100
149           5.9          3.0           5.1          1.8  virginica      100

[150 rows x 6 columns]

In [104]:
iris.new_col2 = 200 # iris객체에 new_col2속성이 추가, 열을 추가하려면 []을 사용해야함
iris

sepal_length  sepal_width  petal_length  petal_width    species  new_col
0             5.1          3.5           1.4          0.2     setosa      100
1             4.9          3.0           1.4          0.2     setosa      100
2             4.7          3.2           1.3          0.2     setosa      100
3             4.6          3.1           1.5          0.2     setosa      100
4             5.0          3.6           1.4          0.2     setosa      100
..            ...          ...           ...          ...        ...      ...
145           6.7          3.0           5.2          2.3  virginica      100
146           6.3          2.5           5.0          1.9  virginica      100
147           6.5          3.0           5.2          2.0  virginica      100
148           6.2          3.4           5.4          2.3  virginica      100
149           5.9          3.0           5.1          1.8  virginica      100

[150 rows x 6 columns]

In [105]:
iris.head()

sepal_length  sepal_width  petal_length  petal_width species  new_col
0           5.1          3.5           1.4          0.2  setosa      100
1           4.9          3.0           1.4          0.2  setosa      100
2           4.7          3.2           1.3          0.2  setosa      100
3           4.6          3.1           1.5          0.2  setosa      100
4           5.0          3.6           1.4          0.2  setosa      100

In [106]:
iris.drop(0)

sepal_length  sepal_width  petal_length  petal_width    species  new_col
1             4.9          3.0           1.4          0.2     setosa      100
2             4.7          3.2           1.3          0.2     setosa      100
3             4.6          3.1           1.5          0.2     setosa      100
4             5.0          3.6           1.4          0.2     setosa      100
5             5.4          3.9           1.7          0.4     setosa      100
..            ...          ...           ...          ...        ...      ...
145           6.7          3.0           5.2          2.3  virginica      100
146           6.3          2.5           5.0          1.9  virginica      100
147           6.5          3.0           5.2          2.0  virginica      100
148           6.2          3.4           5.4          2.3  virginica      100
149           5.9          3.0           5.1          1.8  virginica      100

[149 rows x 6 columns]

In [107]:
iris.drop(1, inplace=True)

In [108]:
iris.drop([2,3,4]).head()

sepal_length  sepal_width  petal_length  petal_width species  new_col
0           5.1          3.5           1.4          0.2  setosa      100
5           5.4          3.9           1.7          0.4  setosa      100
6           4.6          3.4           1.4          0.3  setosa      100
7           5.0          3.4           1.5          0.2  setosa      100
8           4.4          2.9           1.4          0.2  setosa      100

In [109]:
iris.drop(1:4, inplace=True)

SyntaxError: invalid syntax (<ipython-input-109-a7554ae27c41>, line 1)

In [110]:
iris = sns.load_dataset('iris')

In [111]:
iris.drop('species', axis=1).head()

sepal_length  sepal_width  petal_length  petal_width
0           5.1          3.5           1.4          0.2
1           4.9          3.0           1.4          0.2
2           4.7          3.2           1.3          0.2
3           4.6          3.1           1.5          0.2
4           5.0          3.6           1.4          0.2

In [112]:
iris.drop(['sepal_length', 'sepal_width'], axis=1).head()

petal_length  petal_width species
0           1.4          0.2  setosa
1           1.4          0.2  setosa
2           1.3          0.2  setosa
3           1.5          0.2  setosa
4           1.4          0.2  setosa

In [113]:
iris.drop(['sepal_length', 'sepal_width'], axis='columns').head()

petal_length  petal_width species
0           1.4          0.2  setosa
1           1.4          0.2  setosa
2           1.3          0.2  setosa
3           1.5          0.2  setosa
4           1.4          0.2  setosa

In [114]:
iris.drop('species',axis=1).head()

sepal_length  sepal_width  petal_length  petal_width
0           5.1          3.5           1.4          0.2
1           4.9          3.0           1.4          0.2
2           4.7          3.2           1.3          0.2
3           4.6          3.1           1.5          0.2
4           5.0          3.6           1.4          0.2

In [115]:
iris.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

In [116]:
iris.drop('species',axis=1, inplace=True)
iris.head()

sepal_length  sepal_width  petal_length  petal_width
0           5.1          3.5           1.4          0.2
1           4.9          3.0           1.4          0.2
2           4.7          3.2           1.3          0.2
3           4.6          3.1           1.5          0.2
4           5.0          3.6           1.4          0.2

In [117]:
iris = sns.load_dataset('iris')

In [118]:
iris['year'] = 2019
iris.head()

sepal_length  sepal_width  petal_length  petal_width species  year
0           5.1          3.5           1.4          0.2  setosa  2019
1           4.9          3.0           1.4          0.2  setosa  2019
2           4.7          3.2           1.3          0.2  setosa  2019
3           4.6          3.1           1.5          0.2  setosa  2019
4           5.0          3.6           1.4          0.2  setosa  2019

In [119]:
iris['no1'] = [1,2,3]

ValueError: Length of values (3) does not match length of index (150)

In [120]:
iris['no2'] = [10, 20, 30] + [None] * 147

In [121]:
iris['no3'] = None
iris.loc[0:2, 'no3'] = [10, 20, 30]
iris.head()

sepal_length  sepal_width  petal_length  petal_width species  year   no2  \
0           5.1          3.5           1.4          0.2  setosa  2019  10.0   
1           4.9          3.0           1.4          0.2  setosa  2019  20.0   
2           4.7          3.2           1.3          0.2  setosa  2019  30.0   
3           4.6          3.1           1.5          0.2  setosa  2019   NaN   
4           5.0          3.6           1.4          0.2  setosa  2019   NaN   

    no3  
0    10  
1    20  
2    30  
3  None  
4  None

In [122]:
iris['no4'] = pd.Series([10, 20, 30], index=[0,1,2])
iris.head()

sepal_length  sepal_width  petal_length  petal_width species  year   no2  \
0           5.1          3.5           1.4          0.2  setosa  2019  10.0   
1           4.9          3.0           1.4          0.2  setosa  2019  20.0   
2           4.7          3.2           1.3          0.2  setosa  2019  30.0   
3           4.6          3.1           1.5          0.2  setosa  2019   NaN   
4           5.0          3.6           1.4          0.2  setosa  2019   NaN   

    no3   no4  
0    10  10.0  
1    20  20.0  
2    30  30.0  
3  None   NaN  
4  None   NaN

In [123]:
iris = sns.load_dataset('iris')

In [124]:
row1 = {'sepal_length' : 10, 'sepal_width' : 5,
        'petal_length' : 20, 'petal_width' : 10,
        'species' : 'setosa'}

In [125]:
new_iris = pd.concat([iris, pd.DataFrame([row1])], ignore_index=True)
new_iris.tail()

sepal_length  sepal_width  petal_length  petal_width    species
146           6.3          2.5           5.0          1.9  virginica
147           6.5          3.0           5.2          2.0  virginica
148           6.2          3.4           5.4          2.3  virginica
149           5.9          3.0           5.1          1.8  virginica
150          10.0          5.0          20.0         10.0     setosa

In [126]:
new_row = pd.Series([1,2,3,4, 'versicolor'], index=iris.columns)
new_row

sepal_length             1
sepal_width              2
petal_length             3
petal_width              4
species         versicolor
dtype: object

In [127]:
new_iris = pd.concat([iris, new_row.to_frame().T], ignore_index=True)

In [128]:
new_iris.tail()

sepal_length sepal_width petal_length petal_width     species
146          6.3         2.5          5.0         1.9   virginica
147          6.5         3.0          5.2         2.0   virginica
148          6.2         3.4          5.4         2.3   virginica
149          5.9         3.0          5.1         1.8   virginica
150            1           2            3           4  versicolor

In [129]:
df1 = pd.DataFrame({'key':['a','b','c','f'], 'c1':[1,2,3,5]})
df2 = pd.DataFrame({'key':['a','b','d','f'], 'c2':[5,6,7,8]})

In [130]:
df1.merge(df2)

key  c1  c2
0   a   1   5
1   b   2   6
2   f   5   8

In [131]:
df1.merge(df2, how='left')

key  c1   c2
0   a   1  5.0
1   b   2  6.0
2   c   3  NaN
3   f   5  8.0

In [132]:
df1.merge(df2, how='right')

key   c1  c2
0   a  1.0   5
1   b  2.0   6
2   d  NaN   7
3   f  5.0   8

In [134]:
df1.merge(df2, how='outer')

key   c1   c2
0   a  1.0  5.0
1   b  2.0  6.0
2   c  3.0  NaN
3   f  5.0  8.0
4   d  NaN  7.0

In [137]:
df3 = pd.DataFrame({'key3':['a','b','c','f'], 'c1':[1,2,3,5]})
df4 = pd.DataFrame({'key4':['a','b','d','f'], 'c2':[5,6,7,8]})

In [138]:
df3.merge(df4, left_on='key3', right_on='key4')

key3  c1 key4  c2
0    a   1    a   5
1    b   2    b   6
2    f   5    f   8

In [141]:
df3.merge(df4, left_index=True, right_index=True)

key3  c1 key4  c2
0    a   1    a   5
1    b   2    b   6
2    c   3    d   7
3    f   5    f   8

In [145]:
df1 = pd.DataFrame({'c1':[1,2,3,4], 'c2':[5,6,7,8]})
df2 = pd.DataFrame({'c3':['a','b','c','d'], 'c4':[1.2, 3.4, 5.5, 7.6]})

In [146]:
pd.concat([df1, df2], axis=1)

c1  c2 c3   c4
0   1   5  a  1.2
1   2   6  b  3.4
2   3   7  c  5.5
3   4   8  d  7.6

In [148]:
df1 = pd.DataFrame({'c1': [1,2], 'c2' : [3,4]})
df2 = pd.DataFrame({'c1': [5,6], 'c4' : [73,8]})

In [149]:
pd.concat([df1, df2], axis=0, sort=False)

c1   c2    c4
0   1  3.0   NaN
1   2  4.0   NaN
0   5  NaN  73.0
1   6  NaN   8.0

In [150]:
df1 = pd.DataFrame({'c1':[1,2,3,4],
                    'c2':[5,6,7,8]}, index=[0,2,4,6])
df2 = pd.DataFrame({'c3':['a','b','c','d'],
                    'c4':[1.2, 3.4, 5.5, 7.6]}, index=[0,1,2,3])

In [151]:
pd.concat([df1,df2], axis=1)

c1   c2   c3   c4
0  1.0  5.0    a  1.2
2  2.0  6.0    c  5.5
4  3.0  7.0  NaN  NaN
6  4.0  8.0  NaN  NaN
1  NaN  NaN    b  3.4
3  NaN  NaN    d  7.6

In [152]:
df1.reset_index(drop=True, inplace=True)
pd.concat([df1, df2], axis=1)

c1  c2 c3   c4
0   1   5  a  1.2
1   2   6  b  3.4
2   3   7  c  5.5
3   4   8  d  7.6

In [153]:
iris = sns.load_dataset('iris')

각 열별로 최솟값 조회

In [154]:
iris.min()

sepal_length       4.3
sepal_width        2.0
petal_length       1.0
petal_width        0.1
species         setosa
dtype: object

각 열별로 최댓값 조회

In [155]:
iris.max()

sepal_length          7.9
sepal_width           4.4
petal_length          6.9
petal_width           2.5
species         virginica
dtype: object

각 열별로 평균값 조회\
| 버전의 차이로 인해\
| iris.mean()로 실행키면 문자열과 구분되지 않아 오류가 발생한다.\
| iris.mean(numeric_only=True)로 매개변수를 추가하여 숫자값에 대해서만 평균을 구하도록 한다.

In [157]:
iris.mean(numeric_only=True)

sepal_length    5.843333
sepal_width     3.057333
petal_length    3.758000
petal_width     1.199333
dtype: float64

각 열별로 중위수 조회\
| 버전의 차이로 인해\
| iris.median()로 실행키면 문자열과 구분되지 않아 오류가 발생한다.\
| iris.median(numeric_only=True)로 매개변수를 추가하여 숫자값에 대해서만 중위수를 구하도록 한다.

In [158]:
iris.median(numeric_only=True)

sepal_length    5.80
sepal_width     3.00
petal_length    4.35
petal_width     1.30
dtype: float64

In [163]:
iris.var(numeric_only=True)

sepal_length    0.685694
sepal_width     0.189979
petal_length    3.116278
petal_width     0.581006
dtype: float64

In [161]:
#iris.iloc[:, :-1] # 독립변수만 뽑기, 마지막 열빼고 뽑기
A = iris.iloc[:, :-1].to_numpy()
np.var(A), A.var()

(3.896056416666667, 3.896056416666667)

In [162]:
np.var(A, axis=0), A.var(axis=0) # 각 열별 분산

(array([0.68112222, 0.18871289, 3.09550267, 0.57713289]),
 array([0.68112222, 0.18871289, 3.09550267, 0.57713289]))

numpy에도 분산이 있다.\
numpy와 pandas간 분산 구하는 방법이 다르다.\
numpy로 구한 분산 값 : 0.68112222\
pandas로 구한 분산 값 : 0.685694\
결과값이 다르게 나온다.\
\
ddof 매개변수 (델타 ~~ )를 1로 두면 같은 결과값이 나온다.

In [164]:
np.var(A, axis=0), A.var(axis=0, ddof=1)

(array([0.68112222, 0.18871289, 3.09550267, 0.57713289]),
 array([0.68569351, 0.18997942, 3.11627785, 0.58100626]))

In [165]:
iris.describe()

sepal_length  sepal_width  petal_length  petal_width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.057333      3.758000     1.199333
std        0.828066     0.435866      1.765298     0.762238
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000

In [166]:
iris.describe(include='all')

sepal_length  sepal_width  petal_length  petal_width species
count     150.000000   150.000000    150.000000   150.000000     150
unique           NaN          NaN           NaN          NaN       3
top              NaN          NaN           NaN          NaN  setosa
freq             NaN          NaN           NaN          NaN      50
mean        5.843333     3.057333      3.758000     1.199333     NaN
std         0.828066     0.435866      1.765298     0.762238     NaN
min         4.300000     2.000000      1.000000     0.100000     NaN
25%         5.100000     2.800000      1.600000     0.300000     NaN
50%         5.800000     3.000000      4.350000     1.300000     NaN
75%         6.400000     3.300000      5.100000     1.800000     NaN
max         7.900000     4.400000      6.900000     2.500000     NaN